searx/plugins/bangs.py

```diff
@@ -0,0 +1,140 @@
+    :param request: The flask request object.
+    :return: A flask response or None if the plugin did nothing
+    """
+
+    search_query = search_query_obj.query
+
+    if search_query == help_bang_operator:
+        return render_help_bangs_page(request)
+    if search_query == show_bangs_operator:
+        return render_available_bangs(request)
+    if is_valid_bang(search_query):
+        available_bangs = search_bangs(search_query)
+        if len(available_bangs) > 0:
+            # If multiple bangs only select the first one
+            bang = available_bangs[0]
+            # TODO add region support.
+            bang_url = bang["regions"]["default"]
+            bang_full_with_user_query = bang_url.replace("{{{term}}}", get_bang_query(search_query))
+            return redirect(bang_full_with_user_query)
+    return None
+
+
+def render_available_bangs(request):
+    return jsonify(_get_bangs_data())
+
+
+def render_help_bangs_page(request):
+    return render_template(
+        "plugins/bangs/help_bangs.html",
+        operators=[
+            {
+                "operator": show_bangs_operator,
+                "description": "Shows all the available bangs on a given instance."
+            },
+            {
+                "operator": help_bang_operator,
+                "description": "Shows this help/setting page."
+            },
+        ],
+    )
+
+
+def get_bang_from_query(raw_query):
+    """
+    Extracts the bang from a search query.
+    :param raw_query: The raw user query coming from the browser
+    :return: If the raw_query is &yt yes yes. It returns yt.
+    """
+
+    bang = raw_query.split(" ")[0]
+    return bang.replace(bang_operator, "")
+
+
+def get_bang_query(raw_query):
+    """
+    Gets the actual user search.
+    :param raw_query: The raw user query coming from the browser
+    :return: For example with the query &yt yes yes then this function will return yes yes.
+    """
+    slitted_raw_query = raw_query.split(" ")
+
+    full_query = ""
+    # Cause the first one is the bang like &yt
+    for raw_query in slitted_raw_query[1:]:
+        full_query += raw_query + "%20"
+    return full_query
+
+
+def is_valid_bang(raw_search_query):
+    """
+    Check whether the given search query is a bang and if it exists in the json bangs_data/bangs.json file.
+    :param raw_search_query: The user his search query in str
+    :return: True if it is a valid bang
+    """
+    return raw_search_query[0] == bang_operator
+
+
+def search_bangs(raw_search_query):
+    """
+    Searches if the bang is available.
+    :param raw_search_query: The search query the user providied
+    :return: Return a list of dicts with all the bangs data (coming from bangs_data/bangs.json)
+    """
+    user_bang = get_bang_from_query(raw_search_query)
+    # Searches for the bang matching the query
+    return list(filter(
+        lambda bang: user_bang in bang["triggers"], _get_bangs_data()
```

Comment:
It would be faster if ```bangs_data``` is dictionary, with the different triggers as a key. The different triggers for the same bang would reference the same structure, so the memory footprint would be the same. All this could be prepare in the ```_get_bangs_data``` function.

Reply:
@dalf Thanks for the feedback! I think that is something easy to change and I will try to implement it.

Reply:
@dalf I updated it to your method which I think is way better.